Arduino/victoria_teensy_ros_node/victoria_teensy_ros_node.ino

```diff
@@ -451,41 +451,68 @@ void doPublishRawImu(void) {
-  static const double RANGE_RAW_RIGHT_BUMPER(MAX_RAW_RIGHT_BUMPER - MIN_RAW_RIGHT_BUMPER);
-  static const double MAX_FORCE_RIGHT_BUMPER(1.0);
+  // These values come from samples taken directly from the robot.
+  // Distances are in meters.
+  static const int BUMPER_BASE_LEFT(620);
+  static const int BUMPER_RAW_SAMPLES_LEFT[NUM_BUMPER_SAMPLES] = { 0, 60, 144, 216, 250, 277, 287 };
+  static const double BUMPER_DISTANCE_SAMPLES_LEFT[NUM_BUMPER_SAMPLES] = { 0, 0.00246, 0.00405, 0.00625, 0.0087, 0.01088, 0.01201 };
+  static const int BUMPER_BASE_RIGHT(629);
+  static const int BUMPER_RAW_SAMPLES_RIGHT[NUM_BUMPER_SAMPLES] = { 0, 85, 172, 238, 265, 285 , 291 };
+  static const double BUMPER_DISTANCE_SAMPLES_RIGHT[NUM_BUMPER_SAMPLES] = { 0, 0.00253, 0.00406, 0.00627, 0.0089, 0.01095, 0.01206 };
   
   ros::Time current_time = ros_nh.now();
 
-  int rawBumperLeft = analogRead(BUMPER_LEFT_PIN);
-  int rawBumperRight = analogRead(BUMPER_RIGHT_PIN);
-
-  double bumperLeft = MAX_FORCE_LEFT_BUMPER 
-                        * ((rawBumperLeft - MIN_RAW_LEFT_BUMPER) / RANGE_RAW_LEFT_BUMPER);
-  double bumperRight = MAX_FORCE_RIGHT_BUMPER 
-                        * ((rawBumperRight - MIN_RAW_RIGHT_BUMPER) / RANGE_RAW_RIGHT_BUMPER);
-
+  // Read the current values, normalize to the base value.
+  int raw_bumper_left = BUMPER_BASE_LEFT - analogRead(BUMPER_LEFT_PIN);
+  int raw_bumper_right = BUMPER_BASE_RIGHT - analogRead(BUMPER_RIGHT_PIN);
+  
+  double bumper_left = calculateBumperDistance(raw_bumper_left, BUMPER_RAW_SAMPLES_LEFT, BUMPER_DISTANCE_SAMPLES_LEFT);
+  double bumper_right = calculateBumperDistance(raw_bumper_right, BUMPER_RAW_SAMPLES_RIGHT, BUMPER_DISTANCE_SAMPLES_RIGHT);
+  
   ros_bumper_left_msg.header.stamp = current_time;
   ros_bumper_left_msg.min_value = 0.0;
-  ros_bumper_left_msg.max_value = MAX_FORCE_LEFT_BUMPER;
-  ros_bumper_left_msg.value = bumperLeft;
+  ros_bumper_left_msg.max_value = BUMPER_DISTANCE_SAMPLES_LEFT[NUM_BUMPER_SAMPLES - 1];
+  ros_bumper_left_msg.displacement = bumper_left;
   
   ros_bumper_right_msg.header.stamp = current_time;
   ros_bumper_right_msg.min_value = 0.0;
-  ros_bumper_right_msg.max_value = MAX_FORCE_RIGHT_BUMPER;
-  ros_bumper_right_msg.value = bumperRight;
+  ros_bumper_right_msg.max_value = BUMPER_DISTANCE_SAMPLES_RIGHT[NUM_BUMPER_SAMPLES - 1];
+  ros_bumper_right_msg.displacement = bumper_right;
 
   ros_bumper_left_pub.publish(&ros_bumper_left_msg);
   ros_bumper_right_pub.publish(&ros_bumper_right_msg);
 }
 
+/**
+ * Calculate the distance the bumper sensor has moved based on the
+ * given reading and samples of reading and distances.
+ */
+double calculateBumperDistance(int reading, const int* reading_samples, const double* distance_samples) {
+  // Pin the reading to between the min and max sample values
+  reading = min(max(reading_samples[0], reading), reading_samples[NUM_BUMPER_SAMPLES-1]);
+
+  // Find the index of the samples the reading falls into
+  int index = 1;
```

Comment:
Not really for efficiency, but more for readability... instead of doing
```
reading_samples[index]
reading_samples[index - 1]
```
etc, can we do something like
```
reading_upper = reading_samples[index]
reading_lower = reading_samples[index - 1]
```
Which would allow
```
double intermediatePercentage = 
     (reading - reading_lower)  / static_cast<double>(reading_upper - reading_lower);
 +  // Return the total distance the bumper sensor has moved.
 +  return distance_lower + intermediatePercentage * (distance_upper - distance_lower);
```

Which I feel is more readable.